### ISSUE TITLE
Run behaviour for suites and tests using tags

### ISSUE THREAD

**Issue by JesusTheHun** [vitest-dev/vitest#5346]
### Clear and concise description of the problem

I usually organize my tests per domain. Often, one domain can be tested in one file, so it can be picked using the CLI.
When a domain is too big so we try to split it into sub-domains, each being tested within a file.

And sometimes, those sub-domains cannot be tested concurrently, because they would conflict during integration tests with the database for example.

### Suggested solution

Be able to add tags to tasks using _object syntax_ : 

```ts
test('should filter the data correctly', { tags: ['domain1'] }, async () => { /* ... */});
```

Then using the CLI, I can pick or exclude tasks : 

```bash
# Run tests for tasks having the tag "api" or "auth"
vitest --run --tag api --tag auth

# Run tests for tasks having the tag "api" and "auth"
vitest --run --tag api auth

# Pick test files that contain the word "user", exclude the tasks having the tag "api"
vitest user --run --no-tag api

# Prevent tasks with the tag "database" to be run in parallel
vitest --run --tag api --no-tag-parallelism database
```

I realize that because there is an option `-t / --testNamePattern`, the name _tag_ can cause some confusion and maybe there is a better name to be given to this _tag_ system.

### Alternative

Lock mechanism with the database. But vitest could schedule the tasks so they are less likely to block each others.

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide/).
- [X] Check that there isn't already an issue that request the same feature to avoid creating a duplicate.

**Comment by hi-ogawa** [comment]
I'm not sure if you saw this, but it looks like playwright just landed `{ tag: ["@slow"] }` syntax in https://playwright.dev/docs/release-notes#version-142

It seems there was a convention of using `test("whatever @slow", ...)` to put `@...` in test title and actually the filtering part is still relying on the regexp magic to the same filtering option `--grep '@this|@that'`, so they didn't introduce new cli option https://playwright.dev/docs/test-annotations#tag-tests

**Comment by JesusTheHun** [comment]
@hi-ogawa I did see that no. It's interesting to see that they integrated something that grew organically :)
Their CLI option gets quite verbose when you require 3 tags in no particular order, compared to my proposal : 

```bash
# Playwright
npx playwright test --grep (@slow|@fast|@medium-pace)\s+(@slow|@fast|@medium-pace)\s+(@slow|@fast|@medium-pace)
# Vitest
vitest --tag slow fast medium-pace
```

And Playwright's way doesn't solve the parallelism issue.

**Comment by sheremet-va** [comment]
I just want to add that it will not be possible to have `--no-api-parallelism` flag because files are running parallel, not tests, and the tag system exists at runtime - we need to collect tests first, but parallelism is decided before that. Concurrency can still work, but I don't like using tags for this - I would recommend a `describe.concurrent` for these use cases.

**Comment by JesusTheHun** [comment]
It is true that parallelism and even sharding are a challenge.
Parallelism can provide a best effort by using a mutex between workers, and using cache to schedule files during the next run.

**Comment by douglasg14b** [comment]
This would be very helpful in cases where you have separate test suites under the same workspace that need to run disparately. 

Like running `e2e` and `unit` tests from separate commands. You have have 50 packages with tests, listing them all out in the `--project` command is.... less than ideal.

**Comment by AriPerkkio** [comment]
> Like running `e2e` and `unit` tests from separate commands. You have have 50, 100+ packages with tests, listing them all out in the `--project` command is.... less than ideal.

Using tags for this would be much slower than using filename patterns, e.g. have `*.unit.test.ts` and `*.e2e.test.ts`. With tags Vitest would have to import every single test file there are, and check whether tests contain the passed tag. With filename pattern you could simply tell Vitest to import every `*.unit.test.ts` file and run them.

**Comment by douglasg14b** [comment]
@AriPerkkio 

> Using tags for this would be much slower than using filename patterns, e.g. have *.unit.test.ts and *.e2e.test.ts. With tags Vitest would have to import every single test file there are, and check whether tests contain the passed tag

Is vitest not already loading these files during test discovery anyways? Unless I'm using spinning rust loading a few hundred files during discovery to pre-empt a test run is peanuts compared to the cost of the actual test run itself. Unless there is an actual performance problem here I wouldn't use a performance problem as the justification to not use tags.

That said, I would be happy with globs as well.

**Comment by sheremet-va** [comment]
> Is vitest not already loading these files during test discovery anyways?

No, it doesn't. With 4 CPU cores, Vitest will spawn 4 workers. It will then import test files in parallel collecting all tests in only those files and then running them. So there is no global collection phase.

**Comment by sheremet-va** [comment]
The team doesn't think this is a feature we want to have. You can achieve the same thing by having the tag in the test name and using the `-t` argument.

**Comment by amitw-nn** [comment]
@sheremet-va that's a real bummer. I'm sure there are good technical reasons not to support tagging, but putting tags on test names is not a good option, especially if one wants to have multiple tags for some tests (for example, a test might be tagged by the relevant feature, the scope and priority - say `@sanity, @fast, @foo_feature` ) or if one wants to mark an entire suite with a feature tag instead of naming each test with redundant information.

**Comment by TheAlexLichter** [comment]
Note that test tags exist now!

https://github.com/vitest-dev/vitest/pull/9478
https://main.vitest.dev/guide/test-tags.html
